subprojects/plugin-use/src/main/java/org/gradle/plugin/use/internal/DefaultPluginRequestApplicator.java

```diff
@@ -85,63 +88,90 @@ public DefaultPluginRequestApplicator(
             addPluginArtifactRepositories(resolveContext, scriptHandler.getRepositories());
         }
-        List<Result> results = resolvePluginRequests(requests, effectivePluginResolver);
 
         final List<Consumer<PluginManagerInternal>> pluginApplyActions = newLinkedList();
         final Map<Result, PluginImplementation<?>> pluginImplsFromOtherLoaders = newLinkedHashMap();
 
+        List<Result> results = resolvePluginRequests(requests, effectivePluginResolver);
         if (!results.isEmpty()) {
-            for (final Result result : results) {
-                applyPlugin(result.request, result.found.getPluginId(), new Runnable() {
-                    @Override
-                    public void run() {
-                        result.found.execute(new PluginResolveContext() {
-                            @Override
-                            public void addLegacy(PluginId pluginId, Object dependencyNotation) {
-                                pluginApplyActions.add(target -> applyLegacyPlugin(target, result, pluginId));
-                                scriptHandler.addScriptClassPathDependency(dependencyNotation);
-                            }
-
-                            @Override
-                            public void add(PluginImplementation<?> plugin) {
-                                pluginApplyActions.add(target -> applyPlugin(target, result, plugin));
-                            }
-
-                            @Override
-                            public void addFromDifferentLoader(PluginImplementation<?> plugin) {
-                                pluginApplyActions.add(target -> applyPlugin(target, result, plugin));
-                                pluginImplsFromOtherLoaders.put(result, plugin);
-                            }
-                        });
-                        String pluginVersion = result.found.getPluginVersion();
-                        if (pluginVersion != null) {
-                            pluginVersionTracker.setPluginVersionAt(
-                                classLoaderScope,
-                                result.found.getPluginId().getId(),
-                                pluginVersion
-                            );
-                        }
-                    }
-                });
+            applyPlugins(scriptHandler, classLoaderScope, pluginApplyActions, pluginImplsFromOtherLoaders, results);
+        }
+
+        exportBuildLogicClassPathTo(classLoaderScope, scriptHandler.getInstrumentedScriptClassPath());
+
+        if (!autoAppliedPlugins.isEmpty()) {
+            final PluginResolver alreadyOnClasspathIgnoringPluginResolver = new AlreadyOnClasspathIgnoringPluginResolver(effectivePluginResolver,
+                new ClassloaderBackedPluginDescriptorLocator(scriptHandler.getClassLoader())
+            );
+
+            List<Result> autoAppliedPluginResults = resolvePluginRequests(autoAppliedPlugins, alreadyOnClasspathIgnoringPluginResolver);
```

Comment:
Could we here only check for the auto applied plugins whether they are on the classpath or not instead of trying to apply them already? Or am I misunderstanding here something?

Reply:
This is not applying the plugins, it is just resolving them. We could check before and resolve them afterwards, but it will not make any difference since we will check if they are on the classpath before resolving anything.